concourse/steps/release.py

```diff
@@ -23,19 +28,303 @@
 )
 
 
-def rebase(
-    git_helper,
-    upstream_ref,
-):
-    upstream_commit_sha = git_helper.fetch_head(upstream_ref).hexsha
-    git_helper.rebase(commit_ish=upstream_commit_sha)
+class ReleaseContext(object):
+    def set_release_commit_sha(self, release_commit_sha):
+        self._release_commit_sha = release_commit_sha
+
+    def release_commit_sha(self):
+        return self._release_commit_sha
+
+
+class TransactionalStep(object):
+    ''' Provides a framework to implement some action in an atomic fashion.
+
+    Instances of this class are intended to perform exactly one change in their apply-method
+    and provide the corresponding undo-functionality in the revert-method
+    '''
+    def __init__(self, release_context: ReleaseContext):
+        self.release_context = release_context
+
+    def validate(self):
+        pass
+
+    def apply(self):
+        raise NotImplementedError
+
+    def revert(self):
+        pass
+
+    def name(self):
+        pass
+
+
+class ReleaseCommitStep(TransactionalStep):
+    def __init__(
+        self,
+        git_helper:GitHelper,
+        repo_dir: str,
+        release_version: str,
+        repository_version_file_path: str,
+        repository_branch: str,
+        rebase_before_release: bool,
+        release_context: ReleaseContext,
+        release_commit_callback: str=None,
+    ):
+        super().__init__(release_context)
+        self.repository_version_file_path = existing_file(
+            pathlib.Path(repo_dir, repository_version_file_path)
+        )
+        self.git_helper = not_none(git_helper)
+        self.repo_dir = existing_dir(repo_dir)
+        self.repository_branch = repository_branch
+        self.release_version = not_empty(release_version)
+        self.rebase_before_release = rebase_before_release
+        self.release_commit_callback = release_commit_callback
+
+    def name(self):
+        return "Release Commit Creator"
```

Comment:
might it make more sense to choose names that describe the action (e.g. 'create release commit')

Then, we would have error messages like:
"Operation 'create release commit' failed"

(or status messages like "Operation 'create release commit' succeeded")